src/utils.ts

```diff
@@ -1,4 +1,64 @@
-import picomatch from 'picomatch';
+import picomatch, { type Matcher } from 'picomatch';
+
+// #region PARTIAL MATCHER
+export interface PartialMatcherOptions {
+  dot?: boolean;
+  nocase?: boolean;
+}
+
+// the result of over 4 months of figuring stuff out and a LOT of help
+export function getPartialMatcher(patterns: string[], options?: PartialMatcherOptions): Matcher {
+  const regexes = patterns.map(pattern => splitPattern(pattern).map(part => picomatch.makeRe(part, options)));
+  return (input: string) => {
+    // no need to `splitPattern` as this is indeed not a pattern
+    const inputParts = input.split('/');
+    for (let i = 0; i < patterns.length; i++) {
+      const patternParts = splitPattern(patterns[i]);
+      const regex = regexes[i];
+      const minParts = Math.min(inputParts.length, patternParts.length);
+      let j = 0;
+      while (j < minParts) {
+        const part = patternParts[j];
+
+        // handling slashes in parts is very hard, not even fast-glob does it
+        // unlike fast-glob we should return true in this case
+        // for us, better to have a false positive than a false negative here
+        if (part.includes('/')) {
+          return true;
+        }
+
+        const match = regex[j].test(inputParts[j]);
+
+        if (!match) {
+          break;
+        }
+
+        // unlike popular belief, `**` doesn't return true in *all* cases
+        // some examples are when matching it to `.a` with dot: false or `..`
+        // so it needs to match to return early
+        if (part === '**' && match) {
```

Comment:
You checked for a match in the if above and break out the loop. So match should be always truthy here. Hence it can be removed